workflow-runtime-android/src/androidTest/java/com/squareup/workflow1/android/AndroidDispatchersRenderWorkflowInTest.kt

```diff
@@ -0,0 +1,415 @@
+        renderings.collect {
+          if (it == "changed state") {
+            // The rendering we were looking for!
+            expectInOrder(1)
+            completedMutex.unlock()
+          } else {
+            expectInOrder(0)
+            Choreographers.postOnFrameRendered {
+              // We are expecting this to happen last, after we get the rendering!
+              expectInOrder(2)
+            }
+            trigger.emit("changed state")
+          }
+        }
+      }
+
+      completedMutex.lock()
+      collectionJob.cancel()
+    }
+
+  @Test
+  fun all_runtimes_handle_rendering_events_before_next_frame() = runTest {
+    val completedMutex = Mutex(locked = true)
+    val workflow = Workflow.stateful<String, String, SimpleScreen>(
+      initialState = "neverends",
+      render = { renderState ->
+        SimpleScreen(
+          name = renderState,
+          callback = {
+            actionSink.send(action(name = "handleInput") { state = "$state+$state" })
+          }
+        )
+      }
+    )
+
+    val renderings = renderWorkflowIn(
+      workflow = workflow,
+      scope = backgroundScope +
+        Dispatchers.Main.immediate,
+      props = MutableStateFlow(Unit).asStateFlow(),
+      runtimeConfig = runtime.runtimeConfig,
+      workflowTracer = null,
+      interceptors = emptyList()
+    ) {}
+
+    val collectionJob = launch(Dispatchers.Main.immediate) {
+      renderings.collect {
+        if (it.name == "neverends+neverends") {
+          // The rendering we were looking for after the event!
+          expectInOrder(1)
+          completedMutex.unlock()
+        } else {
+          expectInOrder(0)
+          Choreographers.postOnFrameRendered {
+            // This should be happening last!
+            expectInOrder(2)
+          }
+          // First rendering, lets call it.
+          it.callback()
+        }
+      }
+    }
+
+    completedMutex.lock()
+    collectionJob.cancel()
+  }
+}
+
+enum class RuntimeOptions(
+  val runtimeConfig: RuntimeConfig
+) {
+  DEFAULT(RENDER_PER_ACTION),
+  RENDER_ONLY(setOf(RENDER_ONLY_WHEN_STATE_CHANGES)),
+  CONFLATE(setOf(CONFLATE_STALE_RENDERINGS)),
+  STABLE(setOf(STABLE_EVENT_HANDLERS)),
+  RENDER_ONLY_CONFLATE(setOf(CONFLATE_STALE_RENDERINGS, RENDER_ONLY_WHEN_STATE_CHANGES)),
+  RENDER_ONLY_PARTIAL(setOf(RENDER_ONLY_WHEN_STATE_CHANGES, PARTIAL_TREE_RENDERING)),
+  RENDER_ONLY_CONFLATE_STABLE(
+    setOf(CONFLATE_STALE_RENDERINGS, RENDER_ONLY_WHEN_STATE_CHANGES, STABLE_EVENT_HANDLERS)
+  ),
+  RENDER_ONLY_PARTIAL_STABLE(
+    setOf(RENDER_ONLY_WHEN_STATE_CHANGES, PARTIAL_TREE_RENDERING, STABLE_EVENT_HANDLERS)
+  ),
+  RENDER_ONLY_CONFLATE_PARTIAL(
+    setOf(CONFLATE_STALE_RENDERINGS, RENDER_ONLY_WHEN_STATE_CHANGES, PARTIAL_TREE_RENDERING)
+  ),
+  RENDER_ONLY_CONFLATE_PARTIAL_STABLE(
+    setOf(
+      CONFLATE_STALE_RENDERINGS,
+      RENDER_ONLY_WHEN_STATE_CHANGES,
+      PARTIAL_TREE_RENDERING,
+      STABLE_EVENT_HANDLERS
+    )
+  )
```

Comment:
Holy crap, do we really need to test _all_ these combinations? Are we running (or planning to) all of these in production??